### DISCUSSION TITLE
new plugin pin extractor

### DISCUSSION THREAD

**Discussion by cleb40** [cedricp/ddt4all#1294] [General]
Hello , i made a new plugin wich is a generic version of the key_programming plugin
but only to calculate PIN and validate it with subdirs for each algorythms ( though i only have 3 as of now  ),
I made my own branch and i want to push it for review, but i think i don't have the rights to do so.
Can you give me the right to push thanks.
Here some screenshots :
![Screenshot from 2024-11-14 19-04-50](https://github.com/user-attachments/assets/c233c1e8-b7d7-46e0-ad10-7a955a21ab90)
![Screenshot from 2024-11-14 19-04-00](https://github.com/user-attachments/assets/bab340a4-e098-4271-b08c-162f12847caf)

**Comment by Furtif** [comment]
> Hello , i made a new plugin wich is a generic version of the key_programming plugin but only to calculate PIN and validate it with subdirs for each algorythms ( though i only have 3 as of now ), I made my own branch and i want to push it for review, but i think i don't have the rights to do so. Can you give me the right to push thanks. Here some screenshots : ![Screenshot from 2024-11-14 19-04-50](https://private-user-images.githubusercontent.com/187409819/386316646-c233c1e8-b7d7-46e0-ad10-7a955a21ab90.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE2MDc5NzUsIm5iZiI6MTczMTYwNzY3NSwicGF0aCI6Ii8xODc0MDk4MTkvMzg2MzE2NjQ2LWMyMzNjMWU4LWI3ZDctNDZlMC1hZDEwLTdhOTU1YTIxYWI5MC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTE0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTExNFQxODA3NTVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lZWFhZmU4Y2I5MWM3OTAyMDZhM2QxNzU5NGIyODU4YmJjMzgxMDc4YTY4YmE3YTM4Yzc5MWVjOGUxZjVkMjhmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.vZ1mOUhU4OF6XdSuD-YensRuSrSz34NLp0ecREHjXos) ![Screenshot from 2024-11-14 19-04-00](https://private-user-images.githubusercontent.com/187409819/386316653-bab340a4-e098-4271-b08c-162f12847caf.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE2MDc5NzUsIm5iZiI6MTczMTYwNzY3NSwicGF0aCI6Ii8xODc0MDk4MTkvMzg2MzE2NjUzLWJhYjM0MGE0LWUwOTgtNDI3MS1iMDhjLTE2MmYxMjg0N2NhZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTE0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTExNFQxODA3NTVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lNGU4NjMwMTVhY2Y0OTQwOTc5ZmZiMjZiMDBjMWI4NDk0ZDQwYWZhZWQyZWRmM2M1NmMxNjEwOWM3YWM1OWE3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.KcGnlROTA4lq7RNpSoJfKl6x7RBn6p0hnz3b9KG6v-8)

cool propose plugin in repo https://github.com/cedricp/ddtplugins

**Comment by cedricp** [comment]
Great, I'll merge that ASAP! How did you find the algos ?

**Reply by Furtif** [reply]
i think pr is not make

**Reply by Furtif** [reply]
set me right to merge in plugins repos also if you want @cedricp

**Reply by cleb40** [reply]
Well , the first algo is from the card_programming plugin, the second i stumbled on it while doing research on internet and the third is actualy the same as the second but with byte order inverted, i found out by doing testing on a pin extractor program ( well known ) i found too
However it has not been tested, and it's an alfa version, it's more to create a structure , don't expect much

Furtif you tell me to propose plugin in repo, what do you mean ? i tried to push it like this : 

git push --set-upstream https://github.com/cedricp/ddtplugins.git cleb40_pinextractor_experimental

but it fails on authentification
i've never used github, so i  don't know the in and outs

I wouldn't advise to merge merge in master just now, people who know what they are doing should test it and maybe update configs

**Reply by Furtif** [reply]
normal git push not ok you need create a pull request

**Reply by Furtif** [reply]
- fork repo https://github.com/cedricp/ddtplugins/fork
- push plugin in you fork
- propose (create a pull request)

**Reply by cleb40** [reply]
i m on it

**Reply by cleb40** [reply]
done..i think

**Reply by cleb40** [reply]
How did you find the Megane 2 algorythm by the way ?

**Comment by cleb40** [comment]
And the logic might not be perfect, it was one algrythm  and commands for a set of ECUs ,
but it seems the commands ( to start diag session , read isk , send APV , read status) , varies a lot in naming between json files for just slight different versions of ECU, so might need to rethink that

**Comment by cleb40** [comment]
i m going to try it on my megane 2 phase 2 as soon as i find the courage ( i need my car and my last key still works more or less )

**Comment by cleb40** [comment]
And i m on linux by the way with python 3.12.2, no tell how it will behave on windows

**Reply by Furtif** [reply]
ok ok need test on windows

**Comment by cleb40** [comment]
FYI , i tested with my car ( after a few patches , because obviously the car_programming plugin i based mine on is not up to date )
When i send the APV Code i get the response : WRONG RESPONSE : NR: Conditions Not Correct Or Request Sequence Error(7F3B22)
Either the algorythm doesn't work for megane 2 phase 2 or there is a more complex procedure to follow than read isk and send APV
From what i read the UCH i select is not a perfect match and so doesn't work

**Comment by Furtif** [comment]
ok i have pr it https://github.com/cedricp/ddtplugins/pull/7

**Comment by Furtif** [comment]
You said you were afraid, if we're going to see this, it's all wrong

**Comment by Furtif** [comment]
@cedricp want to judge

**Comment by cleb40** [comment]
Hi,

I'm surprised to get news from that , i thought you were not
interrested anyway.
And it doesn't work for my car ( with encrypted ISK ).

I did made a few improvments before stoping working on it :
1 or 2 other algorythms from renault pin extrator , ability to unfilter
algorythms if you want to try others , recoding the init_ecu  with the
original function of init_can or init_iso, and a slight better
interface, some algorythms have been simplifided with the ror function
of BitString too.
I tested it on my car , even if i don't have the needed algorythm

It is algorythms for old cars anyway , the new cars have cryptographic
systems, so i'm not sure it is worth it....

Just now, i can't work on it at all not even review my code , because i
try to stop smoking, so i can't concentrate on things too long or it
make me want to smoke ( just writing this mail is too much actually :(
).
But if you decide to go with it i will provide my new code , one way or
another when i have settled...

here the new look


what do you think ?

**Reply by Furtif** [reply]
I don't think anything, I was just surprised by your first analysis, you did the pr then you were scared, now if we all thought like that ddt4all wouldn't exist, if you want I'll close the copy and you put yours back in, I listen

**Reply by Furtif** [reply]
All of this is up to you to decide, participate, or leave the copy, or @cedricp will judge and he is the boss of this

**Reply by Furtif** [reply]
If you come back to your idea and fear, I will close this one and open yours, (in fact, what this one is yours, it just has my name on it, and I don't like usurping anyone's name) but I think you should do it again. , even if your name is in the commits, the copy I made for me is not 100% because the true creator is you

**Reply by Furtif** [reply]
I would prefer it to be your own and avoid copying, and give the merit to those who deserve it

**Comment by cleb40** [comment]
If you re not french just tell me i send it again in english...

Je crois qu'on est tous français , je vais parler en français donc.
Publie la copie si tu veux, moi c'était juste pour dépanner ma voiture
que j'ai fait ça, mes sources sont libres de droit :)
Quand j'ai commencé ça je n'ai pas considéré que publier ces algo
pouvaient donner lieu a des problèmes, je me suis laissé emporté on va
dire. Je ne penses pas que ça pose réellement problème de toute façon
Et comme je n'avais pas de retour, j'ai pensé que vous n'étiez pas
convaincu par le plugin de toute façon, comme j'ai dit il y a des
remaniements a faire.

Comme j'ai fait des modifs depuis et testé sur ma voiture, je disais
juste que je proposerai les mises a jour que j'ai faite en m'appuyant
sur ton push quand j'aurai la disponibilité d'esprit.
Pour l'instant je suis "out" sur le sujet, donc fait ce que tu veux
avec le code.
Cordialement.

**Reply by Furtif** [reply]
LOL tu es Français et je me fais un effort considérable en English obfuscated  MDRRRRRRRRRRRR .... non c'est ton code, je ne veut pas tes mérites je préfère que tu reopen ton truc, j'ai juste fait sa car tas mis dans que tu avais peur, si on va par la tout dd4all est KO

**Comment by cleb40** [comment]
I honestly don't care

**Comment by cleb40** [comment]
Comme j'ai dit j'essaye d'arrêter de fumer et rien que cette
conversation me donne envie d'une cigarette :)
Je veux pas m'occuper de ça pour l'instant ( a 12 balles le paquet
chaque jour, le calcul est vite fait )
J'aurai pas du répondre du tout, fait comme si j'avais rien dit, et
fait toi plaisir si tu veux le publier.

**Reply by Furtif** [reply]
désole :| on dirait que sa ta travailler la notion de censure ! Pour le reste je ne me fais pas plaisir sauf tu est d'accord

**Comment by cleb40** [comment]
Vas y :)
Je te joint un diff ( en tarball ) de mes dernier dev, a toi de voir si
tu veux les integrer ou non

**Reply by Furtif** [reply]
oui si tu veut pas réouvrir ton pr et tu préfère celui open renvoi dans https://github.com/Furtif/ddtplugins/tree/copy_pineextractor

**Reply by Furtif** [reply]
il se rajoutera au pr open
